packages/loopring_v3/contracts/test/AmmPool.sol

```diff
@@ -0,0 +1,461 @@
+        }
+
+        require(ready, "FUNDS_STILL_IN_EXCHANGE");
+
+        // TODO: withdraw proportional to the pool amount owned
+    }
+
+    struct AuxiliaryData
+    {
+        uint numItems;
+    }
+
+    // Processes work in the queue. Can only be called by the exchange owner
+    // before the blocks containing work for this pool are submitted.
+    // This just verifies if the work is done correctly and only then approves
+    // the L2 transactions that were already included in the block.
+    // Uses synchronized logic on L1/L2 to make the onchain logic easy and efficient.
+    function onSubmitBlocks(
+        ExchangeData.Block[] memory blocks,
+        uint                        blockIdx,
+        uint                        txIdx,
+        bytes                memory auxiliaryData
+        )
+        public
+        override
+        online
+        onlyExchangeOwner
+        returns (uint numTransactionsConsumed)
+    {
+        AuxiliaryData memory auxData = abi.decode(auxiliaryData, (AuxiliaryData));
+        ExchangeData.Block memory _block = blocks[blockIdx];
+        bytes32 exchangeDomainSeparator = exchange.getDomainSeparator();
+
+        // First always update the AMM parameters, which also pulls the AMM balances onchain
+        uint[] memory initialAmmBalances = new uint[](tokens.length);
+        uint[] memory ammBalances = new uint[](tokens.length);
+        for (uint i = 0; i < tokens.length; i++) {
+            // Check that the AMM update in the block matches the expected update
+            AmmUpdateTransaction.AmmUpdate memory update = _block.readAmmUpdate(txIdx++);
+            require(update.owner == address(this), "INVALID_TX_DATA");
+            require(update.accountID == accountID, "INVALID_TX_DATA");
+            require(update.tokenID == tokens[i].tokenID, "INVALID_TX_DATA");
+            require(update.feeBips == feeBips, "INVALID_TX_DATA");
+            require(update.tokenWeight == tokens[i].weight, "INVALID_TX_DATA");
+            // Now approve this AMM update
+            update.validUntil = 0xffffffff;
+            bytes32 txHash = AmmUpdateTransaction.hashTx(exchangeDomainSeparator, update);
+            exchange.approveTransaction(address(this), txHash);
+            numTransactionsConsumed++;
+            // AMM account balance now available onchain
+            initialAmmBalances[i] = update.balance;
+            ammBalances[i] = initialAmmBalances[i];
+        }
+
+        // Process work in the queue
+        for (uint n = queuePos; n < queuePos + auxData.numItems; n++) {
+            QueueItem memory item = queue[n];
+            // TODO: all the necessary pool logic, mint/burn LP tokens,...
+            if (item.workType == WorkType.JOIN) {
+                PoolJoin memory join = item.join;
+                for (uint i = 0; i < tokens.length; i++) {
+                    uint amount = join.maxAmountsIn[i];
+                    ammBalances[i] = ammBalances[i].add(amount);
```

Comment:
I think the logic here is incomplete: we need to calculate for each Join/Exit exactly how much token can be added to/removed from ammBalances[i]/locked[i] based on the new layer-2 AMM balances.

Reply:
Yep for sure, lots of things missing here. Amount of funds actually deposited/withdrawn, mint/burn LP tokens based on the amounts deposited/withdrawn, checking if the LP actually wants to deposit/withdraw at the current balances in the account, if the number is worse than expected don't deposit/withdraw anything at all, etc...

Reply:
Ok, just make sure what's needed to be done for a complete impl.